File: mobile/packages/expo-mobile-web-shell/ios/MobileWebActivationMetadata.swift
import Foundation

struct MobileWebActivationRecord: Codable, Equatable {
  let active: String
  let previous: String?
}

func parseMobileWebActivationRecord(_ data: Data) -> MobileWebActivationRecord? {
  guard
    let json = String(data: data, encoding: .utf8),
    isExactMobileWebJsonDocument(json),
    let value = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
    Set(value.keys) == Set(["active"]) || Set(value.keys) == Set(["active", "previous"]),
    let active = value["active"] as? String,
    isMobileWebSha256(active)
  else {
    return nil
  }
  let previous: String?
  if value.keys.contains("previous") {
    guard
      let candidate = value["previous"] as? String,
      isMobileWebSha256(candidate),
      candidate != active
    else {
      return nil
    }
    previous = candidate
  } else {
    previous = nil
  }
  return MobileWebActivationRecord(active: active, previous: previous)
}

File: mobile/packages/expo-mobile-web-shell/android/src/main/java/expo/modules/mobilewebshell/MobileWebActivationMetadata.kt
package expo.modules.mobilewebshell

import org.json.JSONObject
import org.json.JSONTokener

internal data class MobileWebActivationRecord(
  val active: String,
  val previous: String?
)

internal fun parseMobileWebActivationRecord(json: String): MobileWebActivationRecord {
  val value = try {
    require(isExactMobileWebJsonDocument(json))
    val tokens = JSONTokener(json)
    val candidate = tokens.nextValue()
    require(candidate is JSONObject && tokens.nextClean() == '\u0000')
    candidate
  } catch (_: Exception) {
    throw invalidActivation()
  }
  val keys = value.keys().asSequence().toSet()
  require(keys == setOf("active") || keys == setOf("active", "previous")) {
    "mobile_web_activation_invalid"
  }
  val active = value.opt("active") as? String ?: throw invalidActivation()
  val previous = if ("previous" in keys) {
    value.opt("previous") as? String ?: throw invalidActivation()
  } else {
    null
  }
  require(
    isMobileWebSha256(active) &&
      (previous == null || isMobileWebSha256(previous) && previous != active)
  ) {
    "mobile_web_activation_invalid"
  }
  return MobileWebActivationRecord(active, previous)
}

private fun invalidActivation(): IllegalArgumentException =
  IllegalArgumentException("mobile_web_activation_invalid")
